### ISSUE TITLE
if go_backward == 1,  the output seqences should be reversed

### ISSUE THREAD

**Issue by ypxie** [fchollet/keras#1703]
I think there is a bug in the current theano_backend.rnn

**Comment by tboquet** [comment]
Hi, is this the same issue mentionned here : #1674 ?

**Comment by pasky** [comment]
...and also #1431.  I'm personally carrying a local patch to reverse go_backward==1 for a long time now.  If someone had time to contribute an example of Lambda layer that we could add to RNN's documentation, we could do at least that before the Bidirectional() wrapper is ready, I think it could help a lot of people...

**Comment by shampool** [comment]
Thanks for referring those issues.  I reversed the output in the theaano-backend.  This is actually the way the last released Keras used to handle this issue.

**Comment by shampool** [comment]
Api wise. I think the output should always be in the same order to the input sequence.  If we really need the reverse output. We can explicitly reverse the input.  Otherwise it is very misleading especially without documentations concerning this.

**Comment by pasky** [comment]
I agree with that.  This may be influenced by the fact that I have no prior Theano experience though, so I'm not thinking of how scan() behaves when using the Keras RNN API.

**Comment by farizrahman4u** [comment]
Output reversal should be done at the model level. Anything other than that (such as reversing in `Recurrent.get_output()` or `K.rnn()`) would be too implicit and counter intuitive.

**Comment by shampool** [comment]
This will make thing complicated.  Especially when working on multi dimensional RNN.

**Comment by shampool** [comment]
It is very nature to think about getting outputs that have exactly the same order with the input from recurrent layer.  Keras is a high level Api, not scan level.

**Comment by pasky** [comment]
@farizrahman, I think the dichotomy here is that it'd be too implicit
and counter intuitive for the most experienced people, especially who
are used to work on Theano level, but the current behavior is surprising
and counter intuitive for people who see *only* the Recurrent API and
are just thinking about making outputs depend on future rather than past
time steps.

What about making the go_backwards switch a tri-state?  0/None/False,
True/"backward_seq" for outputting the sequence backwards as now,
"forward_seq" for outputting the sequence in the forward order?
I can make a PR for that if it sounds reasonable.

**Comment by farizrahman4u** [comment]
If you are already into heavy stuff like Multi-dim RNN's, there should be no reason why you would have trouble with something like sequence reversal. `K.rnn()` makes perfect sense (to me) as of now. Its the loop function of Keras. It does `for(i=0;i<n;i++)..` normally, and when `go_backwards=True` it does `for(i=n-1;i>=0;i--)..`. 

Also an RNN with `go_backwards=True` should behave differently from one with `go_backwards=False`. Consider the following RNN which simply copies its input to output:

```python
def step(x, states):
    return x, states
```

If we were to reverse the output as you suggest, the above RNN would behave the same way irrespective of `go_backwards` argument.

**Comment by pasky** [comment]
@farizrahman, I agree completely that this does not belong to the K.rnn API.
My proposal had been only for the Recurrent API.  I'm pretty sure that
whoever is using the K.rnn API already knows what they are doing, but
people using the Recurrent API have a more abstract view of the model.

Basically, just

````python
if self.return_sequences:
    if self.go_backwards == 'forward_seq':
        return outputs[:, ::-1]
    else:
        return outputs
````

in Recurrent.get_output().

**Comment by farizrahman4u** [comment]
@fchollet On second thought, I think @pasky has a point, this is something that should be abstracted by Keras, because its not just that the sequence that has to be reversed, but the mask too! Which can not be done with a `Lambda` layer, so the user is forced to write a custom layer which reverse both the sequence and the mask.

**Comment by shampool** [comment]
@farizrahman4u  I have no trouble with that, I am just saying it is very misleading due to the historical reason and lambda layer along with an extra function complicate things without any gains. 

Why do we need to bother to do those things if that can be easily and naturally handled by just adding the following to K.rnn:

```
    if go_backwards:
       outputs = outputs[::-1]
    outputs = outputs.dimshuffle((1, 0, 2))
```

**Comment by farizrahman4u** [comment]
@fchollet On second thought, I think @pasky has a point, this is something that should be abstracted by Keras, because if the sequence is not reversed, the mask and the sequence will not be aligned.

If we have an adder RNN with `go_backwards=True` and the following sequence is processed:

```
seq : 1, 2, 3, 4, 0, 0, 0
mask: 1, 1, 1, 1, 0, 0, 0
```

The output would be:

```
seq : 0, 0, 0, 4, 7, 9, 10
mask : 1, 1, 1, 1, 0, 0, 0
```

This is very wrong!

**Comment by fchollet** [comment]
Yes, that doesn't seem right. We can handle this in the `Recurrent` layer as suggested a few times before.

**Comment by bryan-lunt** [comment]
Or handle it in the theano_backend. I just submit a patch to fix this.

**Comment by farizrahman4u** [comment]
@bryan-lunt It has to be fixed in `Recurrent` class, not the backend.

**Comment by bryan-lunt** [comment]
@farizrahman4u I guess I don't understand. If it's reversed in the backend, any higher-level recurrence can be written once and automatically benefit from go_backwards. It will simplify writing and debugging any higher level recurrent networks by eliminating special cases there.

**Comment by shampool** [comment]
The backend has been fixed.

**Comment by bryan-lunt** [comment]
@shampool as of when? I grabbed the newest branch a day or so ago and ran some tests with LSTM and SimpleRNN and they never worked backwards.

**Comment by shampool** [comment]
sorry I made a mistake, I just saw your pull request, it's not been merged yet.

**Comment by bryan-lunt** [comment]
OK.
I understand some arguments that "maybe people have already worked around this issue and actually fixing it will break their workarounds." But that's not a very satisfying solution to my mind.

**Comment by bryan-lunt** [comment]
I'm working to put my notebook that demonstrates that my patch works into a gist. I just need to clean it up and comment it. I assume everyone uses Jupyter notebooks? (I'll make a regular .py version too, just in case.)

**Comment by pasky** [comment]
I don't remember seeing such an argument in this issue.  The argument
for doing it in the frontend is that lowlevel Keras hackers used to
Theano will be surprised by the implicit reversal, and it might not
even be what you want for a variety of future non-Recurrent() K.rnn
uses.

(I don't personally feel that these issues are *hugely* important
or convincing, but the core developers clearly do.  So it seems best
to me to take the path of the least resistance. :-)

**Comment by bryan-lunt** [comment]
Won't low-level users used to Theano understand how the `theano.scan` 's `go_backwards` argument works?

**Comment by pasky** [comment]
Yes, and that's the reason why they won't expect the output sequence
to go forward than backwards, as you did in your PR.

**Comment by bryan-lunt** [comment]
OK, I understand now. This is gross.

**Comment by bryan-lunt** [comment]
Here's the promised gist. https://gist.github.com/bryan-lunt/b20b5f15935ab077a400
I'm going to use my own fork of Keras from now on. (And several other people all have private forks now, over this very issue.) All I want is for my science to work.

**Comment by lemuriandezapada** [comment]
The problem is now that if you feed a go_backward layer a sequence [1 2 3 4 5] (that gets processed [5 4 3 2 1]) the output will be returned [e d c b a]. Which is dumb, you'd want it still returned [a b c d e].

That way you can use it in conjuncture with a time_forward layer that also returns it in the [1 2 3 4 5] -> [a b c d e] order for that bidirectional goodness. Otherwise you're aligning them wrong.

If you just read the final output, then you don't care. But if you're actually using the sequence then it's important.

**Comment by bryan-lunt** [comment]
? That's before my patch, but with my patch the output comes out in the right direction. By "now" you mean the current state of the master branch, right?

The maintainer pointed out that I haven't fixed TensorFlow and he did suggest doing it in the `Recurrent` object. I thought doing it at this spot would make it get compiled into Theano functions so that it would benefit from optimizations.

**Comment by farizrahman4u** [comment]
The problem here is that the name `rnn` is misleading, it should have been something like 'loop' or 'scan'.. Its a function for looping.. it doesnt worry about alligning with input and stuff. "nn" stands for neural network, but rnn is a general function, used for many things other than 'RNN's.

**Comment by sunil-at-gh** [comment]
+1 for users carrying their own fork to address this.

I expected SimpleRNN to do the right thing, for me to build a Forward-Backward RNN graph. I solved it by adding a flag (in my copy) to the Theano version of K.rnn() to reverse the returned sequence if that flag and 'go_backward' were both True. In the recurrent.* classes I added that flag in the call.

**Comment by JohnHBrock** [comment]
AFAIK, this is still an issue.
